pkg/api/persistentvolumeclaim/util.go

```diff
@@ -38,6 +38,13 @@ func DropDisabledFields(pvcSpec *core.PersistentVolumeClaimSpec) {
 	if !utilfeature.DefaultFeatureGate.Enabled(features.AnyVolumeDataSource) {
 		pvcSpec.DataSourceRef = nil
 	}
+
+	// Drop the contents of the dataSourceRef field if the CrossNamespaceVolumeDataSource
+	// feature gate is disabled and dataSourceRef.Namespace is specified.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) &&
```

Comment:
I think there is a problem with how this is called from `PrepareForUpdate` (and the AnyVolumeDataSource handling is also problematic). Using https://github.com/kubernetes/kubernetes/blob/edac6fce2ad4a1366a73afc8ac9ca50b2fbf7fc7/pkg/registry/storage/csidriver/strategy.go#L84 as an example, if the old field is set, we should continue allowing it to be set, even if the feature gate is off. The reason for that is explained [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version):

"Ensuring existing data is preserved is needed so that when the feature is enabled by default in a future version n and data is unconditionally allowed to be persisted in the field, an n-1 API server (with the feature still disabled by default) will not drop the data on update."

Reply:
Please see. (https://github.com/kubernetes/kubernetes/pull/113186/commits/9cd22ad1b46c05d8f10f8b6717676f5f977fbc93)
Added an argument to the DropDisabledFields function and added a condition for dropping.